data_safe_haven/infrastructure/components/composite/entra_application.py

```diff
@@ -0,0 +1,158 @@
+
+    def __init__(
+        self,
+        name: str,
+        props: EntraApplicationProps,
+        opts: ResourceOptions | None = None,
+    ) -> None:
+        super().__init__("dsh:common:EntraApplicationComponent", name, {}, opts)
+
+        # Create the application
+        self.application = entra.Application(
+            f"{self._name}_application",
+            display_name=props.application_name,
+            prevent_duplicate_names=True,
+            required_resource_accesses=(
+                [
+                    entra.ApplicationRequiredResourceAccessArgs(
+                        resource_accesses=[
+                            entra.ApplicationRequiredResourceAccessResourceAccessArgs(
+                                id=props.msgraph_permissions[permission_type][
+                                    permission
+                                ],
+                                type=permission_type.value,
+                            )
+                            for permission_type, permission in props.application_permissions
+                        ],
+                        resource_app_id=props.msgraph_client_id,
+                    )
+                ]
+                if props.application_permissions
+                else []
+            ),
+            sign_in_audience="AzureADMyOrg",
+            **props.application_kwargs,
+        )
+
+        # Get the service principal for this application
+        self.application_service_principal = entra.ServicePrincipal(
+            f"{self._name}_application_service_principal",
+            client_id=self.application.client_id,
+        )
+
+        # Grant admin approval for requested application permissions
+        for permission_type, permission in props.application_permissions:
+            if permission_type == EntraAppPermissionType.APPLICATION:
+                entra.AppRoleAssignment(
+                    replace_separators(
+                        f"{self._name}_application_role_grant_{permission_type.value}_{permission}",
+                        "_",
+                    ).lower(),
+                    app_role_id=props.msgraph_permissions[permission_type][permission],
+                    principal_object_id=self.application_service_principal.object_id,
+                    resource_object_id=props.msgraph_object_id,
+                )
+            if permission_type == EntraAppPermissionType.DELEGATED:
+                entra.ServicePrincipalDelegatedPermissionGrant(
+                    replace_separators(
+                        f"{self._name}_application_delegated_grant_{permission_type.value}_{permission}",
+                        "_",
+                    ).lower(),
+                    claim_values=[permission],
+                    resource_service_principal_object_id=props.msgraph_object_id,
+                    service_principal_object_id=self.application_service_principal.object_id,
+                )
```

Comment:
Might be clearer if we can do this as a pair of list comprehensions. Something like

```Python
[entra.AppRoleAssignment(...permission...) for permission_type, permission in ... if permission_type is EntraAppPermissionType.APPLICATION]
[entra.ServicePrincipalDelegatedPermissionGrant(...permission...) for permission_type, permission in ... if permission_type is EntraAppPermissionType.DELEGATED]
```

Reply:
Maybe? It means you need to iterate through the list twice, and I'm not sure whether it's really clearer or not. Happy to do it if you think it's worth it though.

Reply:
Yeah, I doubt this is the rate limiting part of the code though.

_I_ think it is clearer. I've definitely been using list comprehensions long enough that `[something(item) for item in items if something_else(item)]` looks a lot more sensible than a for loop with conditional blocks.

Reply:
Would still like this change unless there is a problem with it.

Reply:
I think replacement would look like:

```
[
            entra.AppRoleAssignment(
                replace_separators(
                    f"{self._name}_application_role_grant_{permission_type.value}_{permission}",
                    "_",
                ).lower(),
                app_role_id=props.msgraph_permissions[permission_type][permission],
                principal_object_id=self.application_service_principal.object_id,
                resource_object_id=props.msgraph_object_id,
            )
            for permission_type, permission in props.application_permissions
            if permission_type == EntraAppPermissionType.APPLICATION
        ]
        [
            entra.ServicePrincipalDelegatedPermissionGrant(
                replace_separators(
                    f"{self._name}_application_delegated_grant_{permission_type.value}_{permission}",
                    "_",
                ).lower(),
                claim_values=[permission],
                resource_service_principal_object_id=props.msgraph_object_id,
                service_principal_object_id=self.application_service_principal.object_id,
            )
            for permission_type, permission in props.application_permissions
            if permission_type == EntraAppPermissionType.DELEGATED
        ]
```

Is this what you were thinking of? I'm not convinced that it's better than what we currently have.

Reply:
Yep, I prefer that.